### ISSUE TITLE
Use BsonReader & BsonWriter

### ISSUE THREAD

**Issue by zigzago** [jershell/kbson#2]
Hello,

Trying to implement a first version of org.bson.codecs.Codec ( https://github.com/Litote/kmongo/commit/8b03c1343ddcbbcfce6c551c09d446bdbe426751#diff-43a261a5b319c5b393c2428d036ce9f4R39 ), I hit a first issue.

encode & decode methods of Codec use BsonWriter & BsonReader but kbson takes BsonDocumentWriter & BsonDocument

Is it doable to change Encoder and Decoder input type?

Thank you

**Comment by jershell** [comment]
Hello,
I will check and try to change it

**Comment by jershell** [comment]
The "BsonWriter" doesn't trouble. The BsonReader it is interface and has implementation Binary and Document. I opinion that needs add both support. What I should make a first?

**Comment by jershell** [comment]
Hello. I finished.
I added BsonDecoder(with BsonReader) and related method the load()
I changed type in encoder on BsonWriter
The method parse() continue to use BsonDocumentDecoder(i renamed to avoid confuse)
The BsonDecoder required strict order of a fields

**Comment by jershell** [comment]
https://github.com/Litote/kmongo/issues/104

**Comment by zigzago** [comment]
@jershell The strict order of fields during deserialization is quite problematic :(

For example, this class does not deserialize well:

```kotlin
@Serializable
data class Friend(
    var name: String?,
    val address: String?,
    @NonEncodeNull
    @ContextualSerialization
    val _id: ObjectId? = null)
```

because in the bson coming from Mongo, the _id comes always first.

But this one is ok (of course):
```kotlin
data class Friend2(
    @NonEncodeNull
    @ContextualSerialization
    val _id: ObjectId? = null,
    var name: String?,
    val address: String?)
```
For the other fields, Mongo returns the fields in the insertion order, so it works. But imagine you insert a document using the shell in a different field order... The deserialization then is also broken.

I would suggest a mechanism similar to the class of the mongo java driver org.bson.codecs.pojo.PojoCodecImpl: 

- Calculate at startup the set of properties to deserialize for the class
- For each deserialization, use a temporary map where you store the field values

- When you read a property during deserialization:
  1) If the property is in the set of properties to deserialize? If not: skip completely the value (perf improvement)
  2) if yes, set the value in the map
- At the end of deserialization, create the instance with the map

**Comment by jershell** [comment]
@zigzago 
I changed BsonDocumentDecoder on work with BsonReader too. It can work with another an order of fields

**Comment by zigzago** [comment]
Thank you! It works fine as you can see: https://github.com/Litote/kmongo/compare/serialization?expand=1

I hit several other issues: #3 #4 #5, the polymorphic one is the biggest - if you fix #3, I think I will be able to release a first version :)
